lib/src/dson.dart

```diff
@@ -51,145 +53,126 @@ class DSON {
+                  receivedType: map.runtimeType.toString(),
+                  parentClass: parentClass,
+                  stackTrace: StackTrace.current,
                 );
-              } else {
-                value = workflow;
               }
-            } else {
-              value = workflow;
-            }
-
-            value = resolvers.fold(
-              value,
-              (previousValue, element) => element(param.name, previousValue),
-            );
-
-            if (value == null) {
-              if (param.isRequired) {
-                if (param.isNullable) {
-                  final entry = MapEntry(
-                    Symbol(param.name),
-                    null,
+
+              final workflow = map[functionParam.aliasOrName];
+
+              if (workflow is Map || workflow is List || workflow is Set) {
+                final innerParam = inner[functionParam.name];
+
+                if (innerParam is IParam) {
+                  value = innerParam.call(
+                    this,
+                    workflow,
+                    inner,
+                    resolvers,
+                    aliases,
                   );
-                  return entry;
-                } else {
-                  throw DSONException(
-                    'Param $className.${param.name} '
-                    'is required and non-nullable.',
+                } else if (innerParam is Function) {
+                  value = fromJson(
+                    workflow,
+                    innerParam,
+                    aliases: aliases,
                   );
+                } else {
+                  value = workflow;
                 }
               } else {
-                return null;
+                value = workflow;
               }
-            }
-
-            final entry = MapEntry(Symbol(param.name), value);
-            return entry;
-          },
-        )
-        .where((entry) => entry != null)
-        .cast<MapEntry<Symbol, dynamic>>()
-        .toList();
 
-    final namedParams = <Symbol, dynamic>{}..addEntries(params);
+              value = resolvers.fold(
+                value,
+                (previousValue, element) =>
+                    element(functionParam.name, previousValue),
+              );
+
+              if (value == null) {
+                if (functionParam.isRequired) {
+                  if (functionParam.isNullable) {
+                    final entry = MapEntry(
+                      Symbol(functionParam.name),
+                      null,
+                    );
+                    return entry;
+                  } else {
+                    throw ParamNullNotAllowed(
+                      functionParam: functionParam,
+                      parentClass: parentClass,
+                      stackTrace: StackTrace.current,
+                    );
+                  }
+                } else {
+                  return null;
+                }
+              }
```

Comment:
Poderia evitar os if's encadeados invertendo a lógica deles:

```suggestion
              if (value = null) {
                if (!functionParam.isRequired) return null;
                if (!functionParam.isNullable) {
                  throw ParamNullNotAllowed(
                    functionParam: functionParam,
                    parentClass: parentClass,
                    stackTrace: StackTrace.current,
                  );
                }  
              
                final entry = MapEntry(Symbol(functionParam.name), null);
                return entry;
              } 
```

Reply:
refatorado e03bbfcf12b3910cf2a821e421f775da30144206